### PR TITLE
Make prediction thread safe.

### PR DESCRIPTION
**Pull request by trivialfis**
As https://github.com/dmlc/xgboost/pull/6645#issuecomment-768359052 fixed a performance issue, the booster object in dask are no longer copied.  So thread safety in predict must be guaranteed.

### CODE REVIEW

**Comment by trivialfis** [comment]
Close https://github.com/dmlc/xgboost/issues/6057

**Comment by hcho3** [comment]
I wonder why this change appears to break the Java package.

**Comment by trivialfis** [comment]
@hcho3 Sorry, to split up the huge PR I have to cherry pick line by line so I just manually copied some code.  Got the thread temp initialization wrong.

**Comment by trivialfis** [comment]
The `dart` model is still not thread safe.  That will have to be done after merging the predictors.

File: src/predictor/gpu_predictor.cu [hunk 1]
```diff
@@ -501,69 +501,72 @@ size_t SharedMemoryBytes(size_t cols, size_t max_shared_memory_bytes) {
...
   }
   void PredictInternal(EllpackDeviceAccessor const& batch,
+                       DeviceModel const& model,
                        HostDeviceVector<bst_float>* out_preds,
-                       size_t batch_offset) {
+                       size_t batch_offset) const {
     const uint32_t BLOCK_THREADS = 256;
     size_t num_rows = batch.n_rows;
     auto GRID_SIZE = static_cast<uint32_t>(common::DivRoundUp(num_rows, BLOCK_THREADS));
+    DeviceModel d_model;
```
Comment (hcho3) [inline src/predictor/gpu_predictor.cu:539]: This local variable is unused.

**Review by hcho3 — APPROVED** [review]
LGTM

File: src/predictor/gpu_predictor.cu (hunk 1 above)
Comment (trivialfis) [reply src/predictor/gpu_predictor.cu]: Good catch!  Removed.

**Pull request merged**